lint/type-guard_checker.go

```diff
@@ -24,63 +24,95 @@ func NewTypeGuardChecker(ctx *Context) *TypeGuardChecker {
 // Check runs type switch checks for f.
 func (c *TypeGuardChecker) Check(f *ast.File) []Warning {
 	c.warnings = c.warnings[:0]
-	// TODO(quasilyte): do recursive analysis. Type switches
-	// may have other type switches inside their clauses that
-	// can trigger warnings as well.
-	for _, decl := range collectFuncDecls(f) {
-		if decl.Body == nil {
-			continue
-		}
-		ast.Inspect(decl.Body, func(x ast.Node) bool {
-			if stmt, ok := x.(*ast.TypeSwitchStmt); ok {
-				c.check(stmt)
-				return false
-			}
-			return true
-		})
-	}
+	inspectFuncBodies(f, c.inspectNode)
 	return c.warnings
 }
 
-func (c *TypeGuardChecker) check(root *ast.TypeSwitchStmt) {
-	if _, ok := root.Assign.(*ast.ExprStmt); !ok {
-		return
+func (c *TypeGuardChecker) inspectNode(x ast.Node) bool {
+	if _, ok := x.(ast.Stmt); !ok {
+		return false
+	}
+	if stmt, ok := x.(*ast.TypeSwitchStmt); ok {
+		c.checkTypeSwitch(stmt)
+	}
+	return true
+}
+
+func (c *TypeGuardChecker) checkTypeSwitch(root *ast.TypeSwitchStmt) {
+	if _, ok := root.Assign.(*ast.AssignStmt); ok {
+		return // Already with type guard
 	}
-	assert, ok := root.Assign.(*ast.ExprStmt).X.(*ast.TypeAssertExpr)
-	if !ok {
-		return
+	// Must be a *ast.ExprStmt then.
+	expr := root.Assign.(*ast.ExprStmt).X.(*ast.TypeAssertExpr).X
+	object := c.ctx.TypesInfo.ObjectOf(c.identOf(expr))
+	if object == nil {
+		return // Give up: can't handle shadowing without object
 	}
 
 	for i, clause := range root.Body.List {
 		clause := clause.(*ast.CaseClause)
 		// Multiple types in a list mean that assert.X will have
-		// a type of interface{} inside body.
+		// a type of interface{} inside clause body.
 		// We are looking for precise type case.
 		if len(clause.List) != 1 {
 			continue
 		}
-		typ := clause.List[0]
+		// Create artifical node just for matching.
+		assert1 := ast.TypeAssertExpr{X: expr, Type: clause.List[0]}
 		for _, stmt := range clause.Body {
-			if c.findTypeAssert(stmt, assert.X, typ) {
+			assert2 := c.findEqual(stmt, &assert1)
+			if object == c.ctx.TypesInfo.ObjectOf(c.identOf(assert2)) {
 				c.warn(root, i)
 				break
 			}
 		}
 	}
 }
 
-func (c *TypeGuardChecker) findTypeAssert(root ast.Stmt, expr, typ ast.Expr) bool {
-	found := false
-	// TODO(quasilyte): inspect does not end the traverse after false is returned.
-	ast.Inspect(root, func(x ast.Node) bool {
-		if assert, ok := x.(*ast.TypeAssertExpr); ok {
-			found = astcmp.EqualExpr(expr, assert.X) &&
-				astcmp.EqualExpr(typ, assert.Type)
-			return !found
+// findEqual walks tree root trying to find node that is equal to x.
+// Matched node is returned.
+func (c *TypeGuardChecker) findEqual(root, x ast.Node) (found ast.Node) {
+	defer func() {
+		if r := recover(); r != nil {
+			panic(r)
```

Comment:
Maybe we could write our own utils to walk through ast and not to use panic? Maybe you should write todo here?

Reply:
> Maybe we could write our own utils to walk through ast

Maybe we could, but that's inefficient time investment right now.
It will require a lot of tests and API design to be usable.
It's simpler to do simple wrappers around `ast` for now, and
think about better design later, when we have more code to consider our needs.

Working with AST is a pain sometimes, but it can't be solved easily.
There are inherent difficulties in it.

> Maybe you should write todo here?

SGTM.

Reply:
There are multiple ways, not even counting making our own traverse functions:
1. Take `golang.org/x/tools/go/ast/astutil` and use it's `Apply`. It's more powerful and has features we want. (Side note, look [how it handles traversal abort](https://github.com/golang/tools/blob/48418e5732e1b1e2a10207c8007a5f959e422f20/go/ast/astutil/rewrite.go#L45-L50)).
2. Use `ast.Walk` with special-purpose visitors that keep `cancel` field inside them and set it to true when certain condition is met.

I'm going to vendor astutil for now.